### PR TITLE
Support for Kubernetes v1.23

### LINKED ISSUE #4823: add chacha20 ciphers to meet DISA STIG compliance
**How to categorize this issue?**
/area security
/kind enhancement
/priority 3
/platform aws-gov

**What would you like to be added**:
Add chacha20 ciphers from k8s v1.23 onwards

**Why is this needed**:
Cipher list was adapted for K8s v1.22 (see [#4300](https://github.com/gardener/gardener/issues/4300) but Gardener's current cipher list for k8s v1.22: 
```
TLS_AES_128_GCM_SHA256
TLS_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
```
is not DISA STIG compliant see [V-242418](https://stigviewer.com/stig/kubernetes/2021-06-17/finding/V-242418)

**What would you like to be added**:
the following ciphers should be added
```
TLS_CHACHA20_POLY1305_SHA256
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305

```

The final TLS 1.3 compliant and DISA STIG compliant cipher list should be (becoming effective with 1.23):
```
TLS_AES_128_GCM_SHA256
TLS_AES_256_GCM_SHA384
TLS_CHACHA20_POLY1305_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
```

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area open-source usability
/kind enhancement
/exp intermediate
/topology garden seed shoot
/merge squash

**What this PR does / why we need it**:
Support for Kubernetes v1.23

**Which issue(s) this PR fixes**:
Part of #5102
Fixes #4823

**Special notes for your reviewer**:
- I didn't vendor new `k8s.io/*` Golang sources yet as we usually do this together with a new `sigs.k8s.io/controller-runtime` upgrade.
- ⚠️ This PR only contains the Gardener part for supporting 1.23 - there will/must be follow-up PRs for each extension repository where individual support gets added.
- Code changes have been tested with the `provider-local`
  - Create/Delete a new shoot cluster with v1.23.1
  - Upgrade an existing cluster to v1.23.1
  - Register a v1.23.1 as a seed cluster

**Release note**:

```feature operator
Gardener can now support shoot and seed clusters with Kubernetes version 1.23. In order to allow creation/update of 1.23 clusters you will have to update the version of your provider extension(s) to a version that supports 1.23 as well. Please consult the respective releases and notes in the provider extension's repository.
```
```feature developer
Gardener can now support seed and shoot clusters with Kubernetes version 1.23. Extension developers have to prepare individual extensions as well to work with 1.23.
```
```noteworthy user
Shoot clusters using Kubernetes 1.23 or above will use the `systemd` cgroup driver as [recommended by the community](https://github.com/kubernetes/kubeadm/issues/1394#issuecomment-462878219).
```

### CODE REVIEW

File: pkg/utils/kubernetes/featuregates.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 // featureGateVersionRanges contains the version ranges for all Kubernetes feature gates.
 // Extracted from https://raw.githubusercontent.com/kubernetes/kubernetes/release-${version}/pkg/features/kube_features.go.
 // To maintain this list for each new Kubernetes version:
-// * Run hack/compare-k8s-feature-gates.sh <old-version> <new-version>.
+// * Run hack/compare-k8s-feature-gates.sh <old-version> <new-version> (e.g. 'hack/compare-k8s-feature-gates.sh v1.22 v1.23').
```
Comment (timebertt) [inline pkg/utils/kubernetes/featuregates.go]:
```suggestion
// * Run hack/compare-k8s-feature-gates.sh <old-version> <new-version> (e.g. 'hack/compare-k8s-feature-gates.sh 1.22 1.23').
```

File: pkg/utils/kubernetes/featuregates.go [hunk 2]
```diff
@@ -39,14 +39,16 @@ var featureGateVersionRanges = map[string]*FeatureGateVersionRange{
...
+	"AllowInsecureBackendProxy":                      {AddedInVersion: "1.17", RemovedInVersion: "1.23"},
 	"AnyVolumeDataSource":                            {AddedInVersion: "1.18"},
 	"AppArmor":                                       {},
 	"AttachVolumeLimit":                              {RemovedInVersion: "1.21"},
 	"BalanceAttachedNodeVolumes":                     {RemovedInVersion: "1.22"},
 	"BlockVolume":                                    {RemovedInVersion: "1.21"},
-	"BoundServiceAccountTokenVolume":                 {},
+	"BoundServiceAccountTokenVolume":                 {RemovedInVersion: "1.23"},
 	"CPUManager":                                     {},
+	"CPUManagerPolicyAlphaOptions":                   {AddedInVersion: "1.22"},
```
Comment (timebertt) [inline pkg/utils/kubernetes/featuregates.go]: Shouldn't all these added features gate have `AddedInVersion=1.23`?

**Review by timebertt — CHANGES_REQUESTED** [review]
Feedback to the first two commits.

File: pkg/utils/kubernetes/featuregates.go (hunk 2 above)
Comment (rfranzke) [reply pkg/utils/kubernetes/featuregates.go]: 🙈 Thanks for catching, sure. Fixed in ea23d09bf.

**Review by timebertt — DISMISSED** [review]
Thank you, looks great! No further comments from my side.
/lgtm

File: charts/istio/istio-ingress/templates/autoscale.yaml [hunk 3]
```diff
@@ -16,4 +16,10 @@ spec:
   - type: Resource
     resource:
       name: cpu
+      {{- if semverCompare "<= 1.22.x" .Capabilities.KubeVersion.GitVersion }}
```
Comment (ialidzhikov) [inline charts/istio/istio-ingress/templates/autoscale.yaml]: Shouldn't the comparison be

```suggestion
      {{- if semverCompare "<= 1.22-0" .Capabilities.KubeVersion.GitVersion }}
```

I recall in the past the `1.22.x` way did not work when Seed is a non Gardener managed cluster and contains a custom version suffix (`v1.19.12-some-suffix`). See https://github.com/gardener/gardener/pull/4676

File: charts/utils-templates/templates/_versions.tpl [hunk 4]
```diff
@@ -43,7 +43,11 @@ batch/v1beta1
 {{- end -}}
 
 {{- define "hpaversion" -}}
+{{- if semverCompare "<= 1.22.x" .Capabilities.KubeVersion.GitVersion -}}
```
Comment (ialidzhikov) [inline charts/utils-templates/templates/_versions.tpl]: Similar to the above question/suggestion - shouldn't it be rather `1.22-0`?

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/config.go [hunk 5]
```diff
@@ -51,7 +51,7 @@ func Config(kubernetesVersion *semver.Version, clusterDNSAddress, clusterDomain
 				CacheUnauthorizedTTL: metav1.Duration{Duration: 30 * time.Second},
 			},
 		},
-		CgroupDriver:                     "cgroupfs",
+		CgroupDriver:                     "systemd",
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/config.go:54]: Doesn't this change of the cgroup driver for v1.23 Shoots deserve a separate release note of type noteworthy/breaking? Would it be useful also to document it somewhere under `docs/`?

File: charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-metrics-scraper.yaml [hunk 6]
```diff
@@ -47,6 +49,11 @@ spec:
           readOnlyRootFilesystem: true
           runAsUser: 1001
           runAsGroup: 2001
+      {{- if semverCompare "> 1.22.x" .Capabilities.KubeVersion.GitVersion -}}
```
Comment (ialidzhikov) [inline charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-metrics-scraper.yaml]: Gets rendered as:
```yaml
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsUser: 1001
          runAsGroup: 2001securityContext:
        seccompProfile:
          type: RuntimeDefault
```

```suggestion
      {{ if semverCompare "> 1.22.x" .Capabilities.KubeVersion.GitVersion -}}
```

File: charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-metrics-scraper.yaml [hunk 7]
```diff
@@ -22,7 +22,9 @@ spec:
         k8s-app: dashboard-metrics-scraper
         origin: gardener
       annotations:
+        {{- if semverCompare "<= 1.22.x" .Capabilities.KubeVersion.GitVersion -}}
```
Comment (ialidzhikov) [inline charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-metrics-scraper.yaml]: Gets rendered as:
```yaml
      annotations:seccomp.security.alpha.kubernetes.io/pod: 'runtime/default'
        # TODO(rfranzke): Remove in a future release.
        security.gardener.cloud/trigger: rollout
```

```suggestion
        {{ if semverCompare "<= 1.22.x" .Capabilities.KubeVersion.GitVersion -}}
```

**Review by ialidzhikov — CHANGES_REQUESTED** [review]
Some minor things that I noticed on my first round of review.

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/config.go (hunk 5 above)
Comment (rfranzke) [reply pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/config.go:54]: Yes, for sure, I've added the release note, thanks. About the `docs`, can you recommend a place where to put this information? So far, we have not documented such details about constraints, so I don't think such document exists yet?

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**